### PR TITLE
Restart services in Docker when changes are made

### PR DESCRIPTION
**Pull request by trushton**
@kevgo @hugobho @martinjaime 
Addresses [Originate/exosphere-sdk#165](https://github.com/Originate/exosphere-sdk/pull/165)
Restarts services affected by file changes, creations, and deletions. The service is stopped, its Docker image rebuilt, and then the service is booted back up in Docker.

~IN-PROGRESS: Feature tests being written for this functionality.~

### CODE REVIEW

File: features/run.feature [hunk 1]
```diff
@@ -37,7 +37,22 @@ Feature: running Exosphere applications
...
 
 
+  Scenario: Editing services of an Exosphere application
+    Given a set-up "running" application
+    When starting "exo-run" in this application's directory
+    Then my machine is running ExoCom
+    And my machine is running the services:
+      | NAME  |
+      | web   |
+      | users |
```
Comment (kevgo) [inline features/run.feature]: I would concatenate all these steps into `Given a running "running" application`

File: features/run.feature [hunk 2]
```diff
@@ -37,7 +37,22 @@ Feature: running Exosphere applications
...
+    When adding a file to the "users" service
+    Then it prints "Restarting service 'users'" in the terminal
+    Then my machine is running the services:
+      | NAME  |
+      | users |
```
Comment (kevgo) [inline features/run.feature]: Why is the web service no longer running?

File: features/run.feature [hunk 3]
```diff
@@ -37,7 +37,22 @@ Feature: running Exosphere applications
...
+    Given a set-up "running" application
+    When starting "exo-run" in this application's directory
+    Then my machine is running ExoCom
+    And my machine is running the services:
+      | NAME  |
+      | web   |
+      | users |
+    When adding a file to the "users" service
+    Then it prints "Restarting service 'users'" in the terminal
+    Then my machine is running the services:
```
Comment (kevgo) [inline features/run.feature]:
```cucumber
Given a running "running" application
When adding a file to the "users" service
Then the "users" service restarts
And the application is still running

File: features/steps/steps.ls [hunk 4]
```diff
@@ -53,6 +54,12 @@ module.exports = ->
     @process.wait expected-text, done
 
 
+  @When /^adding a file to the "([^"]*)" service$/ (service-name) ->
+    app-config = yaml.safe-load fs.read-file-sync(path.join(app-dir, 'application.yml'), 'utf8')
+    service-dir = path.join app-dir, app-config.services[service-name].location
+    child_process.exec-sync "touch #{service-dir}/test.txt"
```
Comment (kevgo) [inline features/steps/steps.ls]: Create a file using `fs.write-file-sync` to avoid going through a subshell.

File: src/service-runner.ls [hunk 5]
```diff
@@ -34,11 +34,44 @@ class ServiceRunner extends EventEmitter
...
 
-    new DockerRunner {@name, @docker-config, @logger}
+    @docker-runner = new DockerRunner {@name, @docker-config, @logger}
         ..start-service!
-        ..on 'online', done
+        ..on 'online', -> done?!
         ..on 'error', (message) ~> @emit 'error', error-message: message
 
+    @watcher = watch @config.root, ignore-initial: yes, ignored: /.*\/node_modules\/.*/
+      ..on 'add', (watch-path) ~>
```
Comment (kevgo) [inline src/service-runner.ls]: `added-path` seems a more descriptive name

File: src/service-runner.ls [hunk 6]
```diff
@@ -34,11 +34,44 @@ class ServiceRunner extends EventEmitter
...
+        @logger.log name: 'exo-run', text: "Restarting service '#{@name}' because #{watch-path} was changed"
+        @restart!
+      ..on 'unlink', (watch-path) ~>
+        @logger.log name: 'exo-run', text: "Restarting service '#{@name}' because #{watch-path} was deleted"
+        @restart!
+
+
+  restart: ->
+    @docker-runner.docker-container.kill!
+    @watcher.close!
```
Comment (kevgo) [inline src/service-runner.ls]: Why are you closing the watcher here? The source code files don't change.

**Comment by trushton** [comment]
All comments addressed.

**Comment by kevgo** [comment]
Tests still failing on Circle with `Error: Command failed: docker rm -f exocom`

File: features/steps/steps.ls [hunk 7]
```diff
@@ -53,6 +52,12 @@ module.exports = ->
     @process.wait expected-text, done
 
 
+  @When /^adding a file to the "([^"]*)" service$/ (service-name) ->
+    app-config = yaml.safe-load fs.read-file-sync(path.join(app-dir, 'application.yml'), 'utf8')
```
Comment (kevgo) [inline features/steps/steps.ls]: Use https://github.com/cutsin/require-yml here to avoid so much boilerplate code

File: features/steps/steps.ls [hunk 8]
```diff
@@ -71,10 +76,22 @@ module.exports = ->
         jsdiff-console actual-routes, expected-routes, done
 
 
-  @Then /^it prints "([^"]*)" in the terminal$/ (expected-text) ->
+  @Then /^it has printed "([^"]*)" in the terminal$/ (expected-text) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: Cool change

File: src/docker-runner.ls [hunk 9]
```diff
@@ -75,11 +75,11 @@ class DockerRunner extends EventEmitter
 
 
   _run-container: ~>
-    new ObservableProcess(@_create-run-command!,
+    @docker-container = new ObservableProcess(@_create-run-command!,
                           stdout: {@write},
```
Comment (kevgo) [inline src/docker-runner.ls]: Please adjust the indentation here

File: src/service-runner.ls [hunk 10]
```diff
@@ -34,11 +34,43 @@ class ServiceRunner extends EventEmitter
...
+        @logger.log name: 'exo-run', text: "Restarting service '#{@name}' because #{removed-path} was deleted"
+        @restart!
+
+
+  restart: ->
+    @docker-runner.docker-container.kill!
+    #TODO: Rebuild the image using exosphere-shared docker-helper
+    service-name = @config.root.split path.sep |> last
+    new ObservableProcess(call-args("docker build --build-arg SERVICE_NAME=#{service-name} -t #{@service-config.author}/#{service-name} ."),
+                         cwd: @config.root,
```
Comment (kevgo) [inline src/service-runner.ls]: Fix indentation

**Comment by trushton** [comment]
@kevgo Tests wont pass on Circle as they still don't allow us to remove containers from the running test environment. Travis has been serving as the point of truth on these tests recently as they allow us to do all of the docker work that we need.

**Comment by kevgo** [comment]
Ah, okay. Let's move off Circle then.

**Pull request merged**